addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js

```diff
@@ -1751,6 +1751,29 @@ export function insertListAfter(afterNode, mode, content = []) {
     return list;
 }
 
+/**
+ * Converts a list element and its nested elements to the specified list mode.
+ *
+ * @param {HTMLElement} node - HTML element representing a list or list item.
+ * @param {string} toMode - Target list mode
+ * @returns {HTMLElement} - Modified element after conversion.
+ */
+export function convertList(node, toMode) {
+    if (!["UL", "OL", "LI"].includes(node.nodeName)) return;
+    const toConvert = closestElement(node, 'ul, ol');
+    const firstChild = toConvert.firstChild;
+    if (toMode !== getListMode(toConvert)) {
+        firstChild.oToggleList(0, toMode);
```

Comment:
Let's say I have a list like
```html
<ul>
  <li>abc</li>
  <li class="oe-nested">
    <ul>
      <li>def</li>
      <li>ghi</li>
    </ul>
  </li>
  <li>jkl</li>
</ul>
```

When pasting this into `<ol><li>[]<br></li>`, the non nested `li` with text abc will be pasted as it is.
However, for the nested `li` with class `oe-nested`, it needs to be converted to an `ol`. So now I have `<ul>` inside the nested li in `convertList`.

If I directly pass the `toConvert` to `oToggleList`, we will have a issue because we do not handle `UL` or `OL` nodes directly in oToggleList see([this](https://github.com/odoo/odoo/blob/16.0/addons/web_editor/static/src/js/editor/odoo-editor/src/commands/toggleList.js)). This handling is delegated to [`HTMLElements.prototype.oToggleList`](https://github.com/odoo/odoo/blob/16.0/addons/web_editor/static/src/js/editor/odoo-editor/src/commands/toggleList.js#L25). Now if it has a [`closestLi`](https://github.com/odoo/odoo/blob/16.0/addons/web_editor/static/src/js/editor/odoo-editor/src/commands/toggleList.js#L29C1-L32C47) here in our case it will be the `li.oe-nested`. It then proceeds to `HTMLLI.prototype.oToggleList`, where it toggles the [closest UL or OL](https://github.com/odoo/odoo/blob/16.0/addons/web_editor/static/src/js/editor/odoo-editor/src/commands/toggleList.js#L70) of the `li.oe-nested` but that's the wrong node to Toggle we needed to toggle the ul inside the `li.oe-nested`. So here by `firstChild` i make sure I always pass a `LI` node to `oToggleList` which toggles the correct list.

Reply:
Thanks for that explanation! So the assumption is that the first child of a nested `UL` or `OL` is _always_ a `LI`? Are you sure that this a reasonable assumption here? This formatting is the one expected in the editor but you're talking about pasting so that's more concerning since outside HTML can have other forms, like `UL > UL`.

Reply:
I have reworked the `convertList` function now it should work for every html list structure(It is tested :green_circle:  in Testing IN stage). I actually ended up moving the logic in `HTMLLIElement.prototype.oToggleList` which converts the list tag to another, to an util function.

Reply:
Great thanks! I made a couple of remarks on the code that seems to be handling that `LI` thing. Have you also addressed the other comments already? If so I'll review the whole PR again, otherwise I'll wait.

Reply:
> Have you also addressed the other comments already? If so I'll review the whole PR again, otherwise I'll wait.

Yes, I have addressed all other comments just the new comment regarding the LI(https://github.com/odoo/odoo/pull/115019#discussion_r1537217007) which is currently in discussion is left to address.